server/src/main/java/com/google/musicanalysis/site/OAuthCallbackServlet.java

```diff
@@ -0,0 +1,76 @@
+package com.google.musicanalysis.site;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse.BodyHandlers;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.JsonParser;
+
+@WebServlet("/api/oauth/callback")
+public class OAuthCallbackServlet extends HttpServlet {
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse res)
+      throws ServletException, IOException {
+    var code = req.getParameter("code");
+    var error = req.getParameter("error");
+    var state = req.getParameter("state");
+
+    // OAuth provider gave us an error
+    if (error != null && !error.isEmpty()) {
+      res.getWriter().print(error);
+      res.setStatus(401);
+      return;
+    }
+
+    // OAuth provider should have given us a code, otherwise something is wrong here
+    if (code == null || code.isEmpty()) {
+      res.setStatus(400);
+      return;
+    }
+
+    var sessionOauthState = req.getSession().getAttribute("oauth:state");
+
+    if (!state.equals(sessionOauthState)) {
+      res.setStatus(401);
+      return;
+    }
+
+    var redirectUri = System.getenv().get("OAUTH_CALLBACK_URI");
```

Comment:
Does this need to be an environment variable? I'm not sure if this is something that needs to be secret like the SPOTIFY_CLIENT_SECRET.

Reply:
I put it in an environment variable since it will be different when we are testing (`http://localhost:8080/api/oauth/callback`) vs. when the app is deployed (`http://step99-2020.appspot.com/api/oauth/callback`).